[WIP] Migration guide for providers to implement V2 primitives

ElePT opened the pull request:
This is a proposal for content to address #2469. After a brief discussion with @jyu00 we agreed we should focus this migration guide on showing how to implement new primitives for providers. However, I have not been involved in all discussions regarding the vision for this migration (and conversations around backend.run tend to raise a lot of different opinions) so I would appreciate **confirmation that the content proposed fits the direction we want to show**.

The page is currently in a pretty raw state, missing formatting, API ref links, a review on the titles, capitalization, etc, but I wanted to open a PR as fast as possible to ask for **opinions on including the backend primitive wrappers as a tool for provider migration**. This idea came from inspecting the only instance I know of external provider beyond the immediate product sphere, which is https://github.com/qiskit-community/qiskit-aqt-provider/tree/master, and subclasses backend primitives to expose their own primitive implementations. Following this, I decided to structure the page as follows:

- motivation: @jyu00 mentioned this was often requested by users, I tried to keep it short and touch on the usefulness of the abstraction level of the primitives 
- Backend primitives as a migration alternative for providers that already implemented backend.run (following the AQT example)
- Custom primitives as an alternative for providers that want it, or didn't implement backend.run (following the StatevectorSampler and StatevectorEstimator) examples

Please let me know if you have any other take on how the content should look like @javabster, @jyu00.

abbycross commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,213 @@
+accessing QPUs. This interface is optimized for two core tasks in quantum algorithm development: expectation value estimation (Estimator) and
+circuit sampling (Sampler).
+
+After a time of co-existence of the `backend.run` model and primitives model, where the primitives
```
```suggestion
After a time of co-existence of the `backend.run` model and primitives model, when the primitives
```
(If I understand correctly, once the primitives evolved to V2, backend.run was deprecated - is that what this is saying?)

abbycross commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,213 @@
+evolved to the updated V2 primitives, the `backend.run` model was finally deprecated in favor of
+these newer interfaces to streamline quantum workload execution.
+
+The existing collection of migration guides helps users [transition to the runtime provider](/migration-guides/qiskit-runtime) and
```
```suggestion
The existing collection of migration guides helps users [transition to the Qiskit Runtime provider](/migration-guides/qiskit-runtime) and
```

abbycross commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,213 @@
+standard Qiskit workflow.
+
+These implementations can be used to wrap any custom provider hardware access function
+(for example: `execute_workload(QPU)` or `resource.access()`), or local simulator,
+as long as the final inputs and outputs conform to the established standards from the primitive interfaces.
```
```suggestion
(for example: `execute_workload(QPU)` or `resource.access()`) or local simulator,
as long as the final inputs and outputs conform to the established standards set by the primitive interfaces.
```

abbycross commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,213 @@
+to the primitives should follow the Primitive Unified Bloc (PUB) syntax specified in the V2 primitives interface. See the [Overview of PUBs section](/guides/primitive-input-output#pubs) in the Primitive inputs and outputs guide for details.
+
+An advantage of this strategy is that the wrapper can handle the input and output manipulation, so knowledge of the PUB data
+model is not required. However, this might result in a sub-optimal runtime, which could be refined through a fully
```
```suggestion
model is not required. However, this might result in a suboptimal runtime, which could be refined through a fully
```

abbycross commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,213 @@
+model is not required. However, this might result in a sub-optimal runtime, which could be refined through a fully
+custom primitives implementation.
+
+The snippets below show how to create a custom Estimator instance following the strategy described above.
```
```suggestion
The following snippets show how to create a custom Estimator instance following the strategy described above.
```

abbycross commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,213 @@
+custom primitives implementation.
+
+The snippets below show how to create a custom Estimator instance following the strategy described above.
+The process would be analogous for a custom Sampler, modifying the base class to `BackendSamplerV2`.
```
```suggestion
The process is analogous for a custom Sampler, modifying the base class to `BackendSamplerV2`.
```

abbycross commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,213 @@
+## If your provider didn't implement `backend.run` or you prefer a fully custom implementation
+
+If a new provider is developed that doesn't conform to the legacy `backend.run` interface, the pre-packaged
+wrapper might not be the optimal route for implementing the primtives. Instead, you should implement a particular
```
```suggestion
wrapper might not be the optimal route for implementing the primitives. Instead, you should implement a particular
```

abbycross commented on docs/migration-guides/external-providers-primitives-v2.mdx:
````diff
@@ -0,0 +1,213 @@
+        )
+```
+
+The mechanics to implement a custom Estimator are analogous to the Sampler, but might require a different pre or post-processing
````
```suggestion
The mechanics to implement a custom Estimator are analogous to those for the Sampler, but might require a different pre- or post-processing
```

abbycross commented on docs/open-source/create-a-provider.mdx:
```diff
@@ -1,14 +1,14 @@
+The Qiskit SDK is built to support third parties in creating external providers of quantum resources.
 
-This means that any organization which develops or deploys quantum compute resources can integrate their services into Qiskit and tap into its userbase. 
+This means that any organization which develops or deploys quantum compute resources can integrate their services into Qiskit and tap into its userbase.
```
```suggestion
This means that any organization that develops or deploys quantum compute resources can integrate their services into Qiskit and tap into its userbase.
```

abbycross commented on docs/open-source/create-a-provider.mdx:
```diff
@@ -1,14 +1,14 @@
+This means that any organization which develops or deploys quantum compute resources can integrate their services into Qiskit and tap into its userbase.
 
-Doing so requires creating a package which supports requests for quantum compute resources and returns them to the user. 
+Doing so requires creating a package which supports requests for quantum compute resources and returns them to the user.
```
```suggestion
Doing so requires creating a package that supports requests for quantum compute resources and returns them to the user.
```

beckykd commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,213 @@
+
+---
+
+# Migrate provider interfaces from backend.run to primitives
```
```suggestion
# Migrate provider interfaces from `backend.run` to primitives
```

ElePT commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,213 @@
+class CustomEstimator(BackendEstimatorV2):
+    """Estimator primitive for custom provider."""
+
+    _backend: CustomProviderResource
```
This line is overwriting the type of the `self._backend` attribute to a custom type, because `BackendEstimatorV2` is written to only accept IBM types. It's a bit of a technicality, Python doesn't enforce these types so the code would run without it, but it feels a bit hacky not to add it. Let me come up with a comment to clarify this.

ElePT replied:
Added in [d1a12e8](https://github.com/Qiskit/documentation/pull/2496/commits/d1a12e80cc756f9fb65e11ea81de83bb41b3f36a)

ElePT commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,216 @@
+
+    # This line is for type checking purposes.
+    # We are changing the type of self._backend from qiskit's
+    # BackendV1/BackendV2 classes our custom provider resource.
```
```suggestion
    # BackendV1/BackendV2 classes to our custom provider resource.
```

jyu00 commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,216 @@
+accessing QPUs. This interface is optimized for two core tasks in quantum algorithm development: expectation value estimation (Estimator) and
+circuit sampling (Sampler).
+
+After a time of co-existence of the `backend.run` model and primitives model, when the primitives
+evolved to the updated V2 primitives, the `backend.run` model was finally deprecated in favor of
+these newer interfaces to streamline quantum workload execution.
```
`backend.run` is not actually deprecated in Qiskit, so we will need to find ways to encourage third party provider to adopt the primitives interface. 

I was thinking something like
> For backward compatibility, the `backend.run` interface continues to exist in Qiskit. However, it is deprecated in Qiskit Runtime, as most of the IBM Quantum users have migrated to primitives due to their improved usability and efficiency. 
> 
> This migration guide aims to help service providers to migrate from the `backend.run` interface to primitives, so their users can also benefits from the improvements.

ElePT replied:
I wasn't sure what level of detail we wanted to share, so I kept it vague to the point where yes, you could infer backend.run is deprecated in Qiskit (I know it's not). I'll apply your suggestion :)

jyu00 commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,216 @@
+evolved to the updated V2 primitives, the `backend.run` model was finally deprecated in favor of
+these newer interfaces to streamline quantum workload execution.
+
+The existing collection of migration guides helps users [transition to the Qiskit Runtime provider](/migration-guides/qiskit-runtime) and
+[update their user code to the V2 primitives interface](/migration-guides/v2-primitives) in time for the upcoming removal of `backend.run`.
+To complement that, this guide supports external providers (both existing and upcoming)
+in implementing their own version of the primitives V2 interface (Estimator or Sampler) that integrates seamlessly into the
+standard Qiskit workflow.
```
I'd put the links to the migration guides in its own section, so the structure becomes 

- why you should implement primitives interface
- how you implement primitives interface
- what you can tell your users

beckykd replied:
Done

jyu00 commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,216 @@
+(for example: `execute_workload(QPU)` or `resource.access()`) or local simulator,
+as long as the final inputs and outputs conform to the established standards set by the primitive interfaces.
+
+## If your provider already implemented `backend.run`
```
I'm a bit confused by this section... this reads like the target audience is people who _use_ 3rd party providers, and not the providers themselves (e.g. users of `IonQProvider` and not IonQ itself).

ElePT replied:
I am not too set on having this section, to be honest, that's why I mentioned in the PR description that I wanted a second opinion on this approach. I based it on the only kind of external (still mostly IBMers) provider that I know implemented primitives, that is: https://github.com/qiskit-community/qiskit-aqt-provider/tree/master. They based their primitives on the backend primitives as shown in this example, and I thought it might be a good shortcut for other providers too, but I also think it might be too hacky. If you don't think it adds any value, it could be removed.

jyu00 replied:
I think we can point out backend primitives as an example, but don't explicitly say they should use it. It's also to their benefit to optimize for their hardware instead of using an out-of-box implementation.

This section would actually be quite helpful in the existing `backend.run` migration guide. If, for example, someone is using `backend.run` across Qiskit Runtime + Provider Foo and want to migrate to primitives. They can use this  as an example to write to primitives interface even if Foo hasn't migrated yet.

jyu00 replied:
I found this page https://docs.quantum.ibm.com/open-source/create-a-provider#providing-an-interface-for-execution, which is a simpler version of what I had in mind. I think we just need to add more "meat" to it, like saying how the workflow is likely 
- validation, such as checking the options are valid
- pre-processing, such as applying any error suppression techniques you support, and broadcasting the inputs 
- execution to get counts
- post-processing, such as calculating the exp val and applying error mit

ElePT commented:
So based on @jyu00's review, it looks like the way forward with this PR will involve the following:

1. Moving the current _If your provider already implemented `backend.run`_ section (L40-93) to [the qiskit runtime backend.run migration guide](https://github.com/Qiskit/documentation/blob/991eeb9a1c14a17e68690089d213ac7970853a53/docs/migration-guides/qiskit-runtime.mdx) and renaming it accordingly. This would address https://github.com/Qiskit/documentation/issues/2543.

2. Keeping the _If your provider didn't implement `backend.run` or you prefer a fully custom implementation_ section as the main example of the guide (renaming it accordingly), and potentially updating the code to match @jyu00's expectations. I took a look at https://docs.quantum.ibm.com/open-source/create-a-provider#providing-an-interface-for-execution and it's essentially a more lightweight version of the example I added. It does look better, but the moment we want to show some useful reference for input/output pre-post processing it gets ugly, so I think that is unavoidable. The classes can of course be renamed. The only element I see in that example compared to the proposed example in this page is the `__init__` method. So I would say the action points are:
  - Adding an init to this example
  - Showing how to do option validation (I think it's the only missing point from this list https://github.com/Qiskit/documentation/pull/2496#discussion_r1901319161)
  
3. As for the introduction/motivation section, the only feedback that hasn't been applied yet is the one in https://github.com/Qiskit/documentation/pull/2496#discussion_r1890780667. I personally like the mention of other migration guides in the introduction, but if you think it will substantially improve the content then I am not so strongly against it. 

If someone from the docs team can take over points 1 and 3, I could try to carve out some time next week to polish the code example from point 2.

ElePT reviewed (commented):
So based on @jyu00's review, it looks like the way forward with this PR will involve the following:

1. Moving the current _If your provider already implemented `backend.run`_ section (L40-93) to [the qiskit runtime backend.run migration guide](https://github.com/Qiskit/documentation/blob/991eeb9a1c14a17e68690089d213ac7970853a53/docs/migration-guides/qiskit-runtime.mdx) and renaming it accordingly. This would address https://github.com/Qiskit/documentation/issues/2543.

2. Keeping the _If your provider didn't implement `backend.run` or you prefer a fully custom implementation_ section as the main example of the guide (renaming it accordingly), and potentially updating the code to match @jyu00's expectations. I took a look at https://docs.quantum.ibm.com/open-source/create-a-provider#providing-an-interface-for-execution and it's essentially a more lightweight version of the example I added. It does look better, but the moment we want to show some useful reference for input/output pre-post processing it gets ugly, so I think that is unavoidable. The classes can of course be renamed. The only element I see in that example compared to the proposed example in this page is the `__init__` method. So I would say the action points are:
  - Adding an init to this example
  - Showing how to do option validation (I think it's the only missing point from this list https://github.com/Qiskit/documentation/pull/2496#discussion_r1901319161)
  
3. As for the introduction/motivation section, the only feedback that hasn't been applied yet is the one in https://github.com/Qiskit/documentation/pull/2496#discussion_r1890780667. I personally like the mention of other migration guides in the introduction, but if you think it will substantially improve the content then I am not so strongly against it. 

If someone from the docs team can take over points 1 and 3, I could try to carve out some time next week to polish the code example from point 2.

abbycross commented:
> If someone from the docs team can take over points 1 and 3, I could try to carve out some time next week to polish the code example from point 2.

Sure we can do that! - FYI @beckykd

beckykd commented:
> 1. Moving the current _If your provider already implemented `backend.run`_ section (L40-93) to [the qiskit runtime backend.run migration guide](https://github.com/Qiskit/documentation/blob/991eeb9a1c14a17e68690089d213ac7970853a53/docs/migration-guides/qiskit-runtime.mdx) and renaming it accordingly.

Done

beckykd commented:
> 3\. As for the introduction/motivation section, the only feedback that hasn't been applied yet

I went ahead and moved the links down.  I think it'll be easier for admins to find and share if it's in a separate section.

ElePT commented:
After a longer pause than initially planned, I have given a push to the content of this PR. The changes include:

1. I have renamed the guides to make it clear that one of them is for backend primitives users (keyword here: backend primitives) in the context of third-party providers, and the other one is the guide to teach third-party provider devs how to implement primitives (keyword: third-party). It was getting confusing for me to tell them apart.
2. As the context on "why migrate to primitives" was moved to the runtime migration overview, I removed the duplicated content in the third-party guide and just reference to that section.
3. I have modified the code examples to adjust to Jessie's comments.  The inspiration is a mix of the statevector primitives and https://docs.quantum.ibm.com/open-source/create-a-provider#providing-an-interface-for-execution.

I believe that the main content concerns have been addressed, so I am requesting a new review :)

ElePT commented:
Note that:

- `docs/migration-guides/qiskit-backend-primitives.mdx` closes https://github.com/Qiskit/documentation/issues/2543
- `docs/migration-guides/external-providers-primitives-v2.mdx` closes https://github.com/Qiskit/documentation/issues/2469

abbycross commented:
@ElePT do you also need another review from @jyu00, or just from myself/@beckykd?

ElePT commented:
I think @jyu00 should take a look if she can. Just to make sure that her comments have been addressed.

abbycross commented on docs/migration-guides/index.mdx:
```diff
@@ -17,15 +17,17 @@ These migration guides help you more effectively use Qiskit and Qiskit Runtime:
    * [Overview](./qiskit-runtime)
    * [Migrate from `qiskit-ibmq-provider`](./qiskit-runtime-from-ibmq-provider)
    * [Migrate from `qiskit_ibm_provider`](./qiskit-runtime-from-ibm-provider)
+   * [Migrate using backend primitives](./qiskit-backend-primitives)
```
I think I'm seeing that the titles aren't consistent for this file. As a general rule, the following must be the same: the title in the metadata, the title in the index.mdx tile, and the title in the _toc.json file. (We sometimes break this rule when titles are inconveniently long for the toc!)

abbycross replied:
```suggestion
   * [Migrate to primitives for users of third-party providers with `backend.run`](./qiskit-backend-primitives)
```

abbycross commented on docs/migration-guides/qiskit-runtime.mdx:
```diff
@@ -53,14 +71,25 @@ See these topics for instructions:
 
-
-### Step 3a: replace any `backend.run` options with `qiskit_ibm_runtime` options.
+### Step 3a: Replace any `backend.run` options with `qiskit_ibm_runtime` options.
```
```suggestion
### Step 3a: Replace any `backend.run` options with `qiskit_ibm_runtime` options
```

abbycross commented on docs/migration-guides/qiskit-runtime.mdx:
```diff
@@ -53,14 +71,25 @@ See these topics for instructions:
+
+## Information to share with your users
+
+There is already a collection of migration guides for users to help with the following tasks:
```
```suggestion
The following migration guides can help users with the following tasks:
```

abbycross commented on docs/migration-guides/qiskit-runtime.mdx:
```diff
@@ -53,14 +71,25 @@ See these topics for instructions:
 - [Migrate options](qiskit-runtime-options)
 - Common use cases [Options section](qiskit-runtime-use-case#migrate-options)
 
+### Step 3b: If you want to access a 3rd-party provider that only exposes `backend.run`, you can still follow a primitive-compatible your workflow using the Qiskit backend primitives
```
I think maybe a word is missing (or is there an extra word?) in this header?

jyu00 replied:
+1 on this. How about

```
Step 3: Replace the call to backend.run with a call to primitives

See these topics for instructions if you are using `qiskit-ibm-runtime`:

- Update code that performs circuit sampling
- Update code that calculates expectation values
- Common use cases (basic, parameterized, and dynamic circuits)
- Replace any backend.run options with Qiskit Runtime primitive options
    - Migrate options
    - Common use cases Options section

See the following topic for instructions if you are using a 3rd-party provider that only exposes backend.run:

...
```

abbycross reviewed (approved):
I made some copyedits directly to the branch, and left a couple questions - but once resolved and once approved by Jessie, it otherwise looks good to me!

jyu00 commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,256 @@
+obtained from sampling from a quantum state in a QPU. The base interfaces provide methods for execution, querying job status, 
+and returning the job results. See the `qiskit.primitives` [API documentation](/api/qiskit/primitives) for more information.
+
+Once you know which primitive to implement, review the PUB data model for input and output handling. See the following 
+guide for more details on [PUBs](/guides/primitive-input-output#pubs). To handle job status and results, 
+the Qiskit SDK provides [`DataBin`](../api/qiskit/qiskit.primitives.DataBin), 
+[`PubResult`](../api/qiskit/qiskit.primitives.PubResult), [`PrimitiveResult`](../api/qiskit/qiskit.primitives.PrimitiveResult), 
+and [`BasePrimitiveJob`](../api/qiskit/qiskit.primitives.BasePrimitiveJob) objects.
```
```suggestion
The Primitive Unified Bloc (PUB) data model defines the input format to the primitives, as described in [Overview of PUBs](https://qiskit.github.io/documentation/pr-2496/guides/primitive-input-output#pubs). While different types of data can be accepted into a PUB, both `SamplerPub` and `EstimatorPub` have a `coerce()` method that can be used to convert them into the respective model. The code snippets below provide an example for each of the primitives.

The [PrimitiveResult](https://qiskit.github.io/documentation/pr-2496/api/qiskit/qiskit.primitives.PrimitiveResult) defines the output of a primitive job. It is comprised of additional data models such as [PubResult](https://qiskit.github.io/documentation/pr-2496/api/qiskit/qiskit.primitives.PubResult) and [DataBin](https://qiskit.github.io/documentation/pr-2496/api/qiskit/qiskit.primitives.DataBin). See [Overview of primitive results](https://docs.quantum.ibm.com/guides/primitive-input-output#overview-of-primitive-results) for more information.
```

jyu00 commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,256 @@
+Once you have migrated your interface to primitives, your users with primitives-based code can seamlessly transition between providers.
+
+
+## Exposing custom `Sampler` and `Estimator` implementations
```
What do you think about changing "custom implementation" to "native implementation"? _Custom_ feels like it's non-standard, but we are asking the providers to implement primitives because they are the new standard.

jyu00 commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,256 @@
+This implementation is based on the [`StatevectorSampler` source code](/api/qiskit/qiskit.primitives.StatevectorSampler),
+with some additions to represent the backend/hardware resource access, as the `StatevectorSampler` class is designed for local statevector simulators.
+
+An example of custom `Estimator` implementation for a third-party provider might look like:
```
I think it'd be more readable if we have separate sections for Sampler and Estimator examples.

jyu00 commented on docs/migration-guides/qiskit-backend-primitives.mdx:
```diff
@@ -0,0 +1,79 @@
+
+# Migrate to primitives for users of third-party providers with `backend.run`
+
+The `backend.run()` interface, previously the primary access point for IBM&reg; QPUs, has been superseded 
+by the Qiskit [primitives interface](/guides/primitives) in `qiskit_ibm_runtime`. 
+While this has become the standard for IBM providers, not all third-party providers have transitioned 
+to the primitives framework, and `backend.run()` remains in Qiskit for backward compatibility.
```
Let's focus this on it being a change in standard interface defined in Qiskit, rather than a change in IBM Quantum.

```suggestion
The `backend.run` interface was the primary access point to quantum hardware and simulators offered by third party providers that integrate with Qiskit. The evolution of [user access patterns for QPUs](/migration-guides/qiskit-runtime#why-migrate-to-qiskit-runtime-primitives) has led to the Qiskit primitives (Sampler and Estimator) superseding this interface. However, not all third-party providers have transitioned to the primitives framework, and `backend.run()` remains in Qiskit for backward compatibility.
```

jyu00 commented on docs/migration-guides/qiskit-backend-primitives.mdx:
```diff
@@ -0,0 +1,79 @@
+[`BackendSamplerV2`.](/api/qiskit/qiskit.primitives.BackendSamplerV2) and 
+[`BackendEstimatorV2`](/api/qiskit/qiskit.primitives.BackendEstimatorV2) is seamless, as
+these classes follow the same usage patterns. The inputs to the primitives should follow 
+the Primitive Unified Bloc (PUB) syntax, and the output formats should follow the
+V2 primitives standard. See the [Overview of PUBs section](/guides/primitive-input-output#pubs) in 
+the Primitive inputs and outputs guide for details.
```
```suggestion
the Primitive Unified Bloc (PUB) syntax, and the output formats should follow the
`PrimitiveResult` interface. See the [Primitive inputs and outputs](/guides/primitive-input-output) for more details.
```

jyu00 commented on docs/migration-guides/qiskit-backend-primitives.mdx:
```diff
@@ -0,0 +1,79 @@
+The two main differences with the implementations of [`SamplerV2`](/api/qiskit-ibm-runtime/sampler-v2) 
+and [`EstimatorV2`](/api/qiskit-ibm-runtime/estimator-v2) are:
+
+1.  The `backend` input passed during the "backend primitive" instantiation. This input argument allows to 
+internally wrap the access to the internal provider through `backend.run()` and it should not be confused
```
```suggestion
1.  The `backend` input passed during the "backend primitive" instantiation. This input argument allows access to the internal provider through `backend.run()` and should not be confused
```

jyu00 commented on docs/migration-guides/qiskit-backend-primitives.mdx:
```diff
@@ -0,0 +1,79 @@
+with the `mode` input argument of [`SamplerV2`](/api/qiskit-ibm-runtime/sampler-v2) 
+and [`EstimatorV2`](/api/qiskit-ibm-runtime/estimator-v2).
+
+2. The use of `options`, which contains one more level of nesting compared to provider-specific primitives and
```
I don't understand where this additional level of nesting comes from, given you can also do `options={"default_shots": 1024}` using qiskit-ibm-runtime.

ElePT replied:
Additional level of nesting is probably not the right term, I wanted to find a nice way to explain that the structure of the options is sometimes different. The level of nesting point referred to `run_options` which may bundle options that are "first level" in the runtime primitives. But I might be confusing things. It might be better to simply lay out how the options look like. The paragraph below does explain that custom option setting might require a subclass anyways.

jyu00 commented on docs/migration-guides/qiskit-backend-primitives.mdx:
```diff
@@ -0,0 +1,79 @@
+[`BackendEstimatorV2`](/api/qiskit/qiskit.primitives.BackendEstimatorV2) interfaces through subclassing to fit the specific 
+interface needs.
+
+The following is an example usage of the [`BackendEstimatorV2`](/api/qiskit/qiskit.primitives.BackendEstimatorV2) 
```
The code uses a sampler, not estimator

```suggestion
The following is an example usage of the [`BackendSamplerV2`](/api/qiskit/qiskit.primitives.BackendSamplerV2) 
```

jyu00 commented on docs/migration-guides/qiskit-backend-primitives.mdx:
```diff
@@ -0,0 +1,79 @@
+tqc = pm.run(qc)
+
+# run using pub syntax
+result = sampler.run(([qc])).result()
```
The additional `()` doesn't do anything. If you want to make it explicitly a tuple, it needs to be `run(([qc],))`

```suggestion
result = sampler.run([qc]).result()
```

jyu00 commented on docs/migration-guides/qiskit-runtime.mdx:
```diff
@@ -17,10 +17,28 @@ examples to help you migrate code from the legacy `backend.run()` interface to u
 </Admonition>
 
-The `qiskit-ibm-runtime` package provides cloud access to the IBM QPUs (quantum processing units) through the primitives interface. The `backend.run()` interface coexisted with the original (V1) primitives model as the dedicated “direct hardware access” entry point.  With the introduction of the V2 primitives interface, the new `SamplerV2` class now fulfills that role. Consequentially, `backend.r ...
+The `qiskit-ibm-runtime` package provides cloud access to the IBM&reg; QPUs (quantum processing units) through the primitives interface. The `backend.run()` interface coexisted with the original (V1) primitives model as the dedicated “direct hardware access” entry point.  With the introduction of the V2 primitives interface, the new `SamplerV2` class now fulfills that role. Consequentially, `back ...
```
```suggestion
The `qiskit-ibm-runtime` package provides cloud access to the IBM&reg; QPUs (quantum processing units) through the primitives interface. The `backend.run()` interface coexisted with the original (V1) primitives model as the dedicated “direct hardware access” entry point.  With the introduction of the V2 primitives interface, the new `SamplerV2` class now fulfills that role. Consequentially, `backend.run()` and `qiskit-ibm-provider`, which only exposed the `backend.run()` interface, are no longer supported.
```

jyu00 commented on docs/migration-guides/qiskit-runtime.mdx:
```diff
@@ -17,10 +17,28 @@ examples to help you migrate code from the legacy `backend.run()` interface to u
+expectation value estimation (Estimator) and circuit sampling (Sampler). The goal is once
+again to help developers to focus more on innovation and less on data conversion.
+
+For backward compatibility, the `backend.run` interface continues to exist in Qiskit. However, it is deprecated
+in Qiskit Runtime, as most IBM Quantum&reg; users have migrated to V2 primitives due to their improved
```
```suggestion
For backward compatibility, the `backend.run` interface continues to exist in Qiskit. However, it is no longer supported
in Qiskit Runtime after the deprecation period, as most IBM Quantum&reg; users have migrated to V2 primitives due to their improved
```

jyu00 commented on docs/migration-guides/qiskit-runtime.mdx:
```diff
@@ -53,14 +71,25 @@ See these topics for instructions:
+
+The following migration guides can help users with the following tasks:
+
+- [Transition to the Qiskit Runtime provider](/migration-guides/qiskit-runtime)
```
Does this just link to the top of this same page?

jyu00 replied:
ok... but why is it called `Transition to the Qiskit Runtime provider`? Is it meant to say transitioning from `qiskit-ibmq-provider` / `qiskit-ibm-provider` to `qiskit-ibm-runtime` ?

ElePT replied:
I'm not sure why it's called like this, but given that the title of the page is now "Migrate to primitives", and that this is self-referencing, I think this one would make more sense:

```suggestion
- [Transition from `backend.run` to primitives](/migration-guides/qiskit-runtime)
```

beckykd replied:
I agree that your suggested title makes more sense.  Currently, the title of this file is actually "Overview", which is clearly a mistake, so I'll update it. You can verify that I named it correctly.

beckykd replied:
Also, @abbycross told me that we're going to use backend.run without (), so I updated those in this file as well.

jyu00 commented on docs/migration-guides/qiskit-runtime.mdx:
```diff
@@ -53,14 +71,25 @@ See these topics for instructions:
+The following migration guides can help users with the following tasks:
+
+- [Transition to the Qiskit Runtime provider](/migration-guides/qiskit-runtime)
+- [Update user code to the V2 primitives interface](/migration-guides/v2-primitives)
```
I'm not sure if this is needed... given it's a migration guide for backend.run to primitives, readers should not have used v1 primitives.

jyu00 reviewed (commented):
Thank you @ElePT for making the changes. I left some suggestions, but I have no faith in my writing so @abbycross please make the editorial review!

abbycross commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,252 @@
+
+## Why should third-party providers implement primitives?
+
+The evolution of [user access patterns for QPUs](/migration-guides/qiskit-runtime#why-migrate-to-qiskit-runtime-primitives?)
```
for anchors, we can usually ignore all punctuation/capitalization so I'll see how it does when we simply remove it.

abbycross commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,252 @@
+
+## Why should third-party providers implement primitives?
+
+The evolution of [user access patterns for QPUs](/migration-guides/qiskit-runtime#why-migrate-to-qiskit-runtime-primitives?)
```
```suggestion
The evolution of [user access patterns for QPUs](/migration-guides/qiskit-runtime#why-migrate-to-qiskit-runtime-primitives)
```

abbycross commented on docs/migration-guides/qiskit-backend-primitives.mdx:
```diff
@@ -6,41 +6,36 @@ description: Migrate to primitives for users of third-party providers with `back
-by the Qiskit [primitives interface](/guides/primitives) in `qiskit_ibm_runtime`. 
-While this has become the standard for IBM providers, not all third-party providers have transitioned 
-to the primitives framework, and `backend.run()` remains in Qiskit for backward compatibility.
+The `backend.run` interface was the primary access point to quantum hardware and simulators offered by third party providers that integrate with Qiskit. The evolution of [user access patterns for QPUs](/migration-guides/qiskit-runtime#why-migrate-to-qiskit-runtime-primitives) has led to the Qiskit primitives (Sampler and Estimator) superseding this interface. However, not all third-party provider ...
```
```suggestion
The `backend.run` interface was the primary access point to quantum hardware and simulators offered by third-party providers that integrate with Qiskit. The evolution of [user access patterns for QPUs](/migration-guides/qiskit-runtime#why-migrate-to-qiskit-runtime-primitives) has led to the Qiskit primitives (Sampler and Estimator) superseding this interface. However, not all third-party providers have transitioned to the primitives framework, and `backend.run()` remains in Qiskit for backward compatibility.
```

abbycross commented on docs/migration-guides/qiskit-runtime.mdx:
```diff
@@ -64,30 +63,24 @@ Follow the steps in the appropriate topic to change your import options and othe
 - [Migrate from `qiskit-ibmq-provider`](qiskit-runtime-from-ibmq-provider)
 
-### Step 3: Replace the call to `backend.run` with a call to `qiskit_ibm_runtime`.
+### Step 3: Replace the call to `backend.run` with a call to primitives.
```
```suggestion
### Step 3: Replace the call to `backend.run` with a call to primitives
```

abbycross commented on docs/migration-guides/qiskit-runtime.mdx:
```diff
@@ -64,30 +63,24 @@ Follow the steps in the appropriate topic to change your import options and othe
-### Step 3b: If you want to access a 3rd-party provider that only exposes `backend.run`, you can still follow a primitive-compatible your workflow using the Qiskit backend primitives.
-
-See the following topic for instructions:
+See the following topic for instructions if you are using a 3rd-party provider that only exposes `backend.run`:
```
```suggestion
See the following topic for instructions if you are using a third-party provider that only exposes `backend.run`:
```

abbycross commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,267 @@
+
+The `Estimator` interface exposes a tool to calculate expectation 
+values of observables with respect to quantum states, while the `Sampler` outputs the counts or probability distributions 
+obtained from sampling from a quantum state in a QPU. These base interfaces also provide standard methods for execution, 
```
"obtained from sampling from a quantum state in a QPU" is a bit awkward - does it change the meaning to say instead, "obtained by sampling a quantum state in a QPU" or something like this? @ElePT

ElePT replied:
absolutely, thanks!

abbycross reviewed (approved):
[One last question](https://github.com/Qiskit/documentation/pull/2496/files#r1989326691) to resolve and then I think it's ready to go

ElePT commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,267 @@
+
+The `Estimator` interface exposes a tool to calculate expectation 
+values of observables with respect to quantum states, while the `Sampler` outputs the counts or probability distributions 
+obtained from sampling from a quantum state in a QPU. These base interfaces also provide standard methods for execution, 
```
```suggestion
obtained by sampling a quantum state in a QPU. These base interfaces also provide standard methods for execution, 
```

jyu00 commented on docs/migration-guides/qiskit-backend-primitives.mdx:
```diff
@@ -0,0 +1,74 @@
+The `run_options` field defines the options to pass through to the `.run()` method of the wrapped backend instance.
+
+Note that because these wrappers are independent of `qiskit_ibm_runtime`, they do not offer access to execution
+modes (`Session`/`batch`), local testing modes, or any other `qiskit_ibm_runtime`-specific feature. These are up to the
```
```suggestion
modes (`Session`/`Batch`), local testing modes, or any other `qiskit_ibm_runtime`-specific feature. These are up to the
```

jyu00 commented on docs/migration-guides/qiskit-runtime.mdx:
```diff
@@ -17,10 +17,28 @@ examples to help you migrate code from the legacy `backend.run()` interface to u
+As access to QPUs became more widespread, and with more quantum algorithms being developed,
+again the need for a higher-level abstraction emerged. In response, IBM introduced
+the Qiskit [primitives interface](/guides/primitives), which are optimized for two core tasks in quantum algorithm development:
+expectation value estimation (Estimator) and circuit sampling (Sampler). The goal is once
```
```suggestion
expectation value estimation (`Estimator`) and circuit sampling (`Sampler`). The goal is once
```

jyu00 commented on docs/migration-guides/qiskit-backend-primitives.mdx:
```diff
@@ -0,0 +1,74 @@
+
+# Migrate to primitives for users of third-party providers with `backend.run`
+
+The `backend.run` interface was the primary access point to quantum hardware and simulators offered by third-party providers that integrate with Qiskit. The evolution of [user access patterns for QPUs](/migration-guides/qiskit-runtime#why-migrate-to-qiskit-runtime-primitives) has led to the Qiskit primitives (Sampler and Estimator) superseding this interface. However, not all third-party provider ...
```
```suggestion
The `backend.run` interface was the primary access point to quantum hardware and simulators offered by third-party providers that integrate with Qiskit. The evolution of [user access patterns for QPUs](/migration-guides/qiskit-runtime#why-migrate-to-qiskit-runtime-primitives) has led to the Qiskit primitives (`Sampler` and `Estimator`) superseding this interface. However, not all third-party providers have transitioned to the primitives framework, and `backend.run` remains in Qiskit for backward compatibility.
```

jyu00 reviewed (approved):
Some minor suggestions for consistency, but otherwise looks good!

jyu00 commented:
@abbycross how come some of the links were added to open new tabs but the rest are not? Is there a standard on what should / should not open a new tab? 

<img width="818" alt="image" src="https://github.com/user-attachments/assets/d7324b04-ac6f-40a1-ad4b-97233ab8d440" />

abbycross commented on docs/migration-guides/external-providers-primitives-v2.mdx:
```diff
@@ -0,0 +1,267 @@
+defines the output of a primitive job. It is comprised of additional data models such as 
+[PubResult](https://qiskit.github.io/documentation/pr-2496/api/qiskit/qiskit.primitives.PubResult) and 
+[DataBin](https://qiskit.github.io/documentation/pr-2496/api/qiskit/qiskit.primitives.DataBin). 
+See [Overview of primitive results](https://docs.quantum.ibm.com/guides/primitive-input-output#overview-of-primitive-results) 
```
```suggestion
See [Overview of primitive results](/guides/primitive-input-output#overview-of-primitive-results) 
```

beckykd reviewed (approved):
Thanks, Elena!

abbycross commented:
[Has this comment been resolved](https://github.com/Qiskit/documentation/pull/2496#discussion_r1989855133)? This PR has gotten too noisy to be absolutely sure :-)

ElePT commented:
@abbycross 

> [Has this comment been resolved](https://github.com/Qiskit/documentation/pull/2496#discussion_r1989855133)? This PR has gotten too noisy to be absolutely sure :-)

Becky modified the page title and I just applied the change to the link text, so I it's been resolved.
